File: src/structs/lotto.rs
use chrono::{DateTime, NaiveDate, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

/// 支援的彩種
pub const GAMES: &[&str] = &["lotto649", "super_lotto638"];
/// 登錄來源
pub const SOURCES: &[&str] = &["qr", "manual"];

/// 一注登錄（DB 對應）
#[derive(Clone, Serialize, FromRow)]
pub struct Ticket {
    pub id: Uuid,
    pub member_id: i64,
    pub game: String,
    pub draw_date: NaiveDate,
    pub picks: Vec<i16>,
    pub second: Option<i16>,
    pub source: String, // 'qr' | 'manual'
    pub checked: bool,
    pub prize_tier: Option<String>,
    pub notified_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// 批次登錄請求：整批共用 game / draw_date / source，notes 帶多注
#[derive(Deserialize)]
pub struct TicketBatchRequest {
    pub game: String,
    pub draw_date: NaiveDate,
    pub source: String,
    pub notes: Vec<NoteInput>,
}

/// 一注的號碼
#[derive(Deserialize)]
pub struct NoteInput {
    pub picks: Vec<i16>,
    pub second: Option<i16>,
}

/// 列表查詢
#[derive(Deserialize)]
pub struct TicketListQuery {
    pub game: Option<String>,
    pub status: Option<String>, // 'pending' | 'won' | 'lost'
    pub page: Option<i64>,
    pub per_page: Option<i64>,
}

/// 通知偏好切換
#[derive(Deserialize)]
pub struct NotifyPrefRequest {
    pub enabled: bool,
}

#[derive(Serialize)]
pub struct NotifyPrefResponse {
    pub enabled: bool,
}

/// 開獎結果（DB 對應，亦作 API 回傳）
#[derive(Clone, Serialize, FromRow)]
pub struct Draw {
    pub game: String,
    pub period: String,
    pub draw_date: NaiveDate,
    pub main_nums: Vec<i16>,
    pub special: i16,
}

/// 開獎查詢
#[derive(Deserialize)]
pub struct DrawListQuery {
    pub game: Option<String>,
    pub limit: Option<i64>,
}

/// 對獎 job 寄信用：一注中獎 + 收件 member email
#[derive(FromRow)]
pub struct WinnerRow {
    pub id: Uuid,
    pub member_id: i64,
    pub game: String,
    pub draw_date: NaiveDate,
    pub prize_tier: String,
    pub email: String,
}

File: frontend/types/lotto.ts
// 大樂透 / 威力彩 選號登錄 + 自動對獎（POST /member/lotto；與發票對獎為不同功能）
export type LottoGame = 'lotto649' | 'super_lotto638';
export type LottoSource = 'qr' | 'manual';
export type LottoStatus = 'pending' | 'won' | 'lost';

// 中獎獎別 key；須搭配 game 解讀（lotto649 到 seventh+general、super_lotto638 到 ninth）
export type LottoPrizeTier =
  | 'first'
  | 'second'
  | 'third'
  | 'fourth'
  | 'fifth'
  | 'sixth'
  | 'seventh'
  | 'eighth'
  | 'ninth'
  | 'general';

// 一注：一般號/第一區 6 個相異；威力彩第二區 second(1-8)，大樂透 second 為 null
export interface LottoNote {
  picks: number[];
  second: number | null;
}

export interface LottoTicket {
  id: string;
  member_id: number;
  game: LottoGame;
  draw_date: string; // 這注要對的開獎日 YYYY-MM-DD
  picks: number[];
  second: number | null;
  source: LottoSource;
  checked: boolean; // 該期是否已開獎並對過
  prize_tier: LottoPrizeTier | null; // checked=true 且為 null = 確定未中
  notified_at: string | null;
  created_at: string;
  updated_at: string;
}

// 批次登錄（POST /member/lotto）：整批共用 game/draw_date/source，notes 帶多注
export interface LottoInput {
  game: LottoGame;
  draw_date: string;
  source: LottoSource;
  notes: LottoNote[];
}

export interface LottoListParams {
  game?: LottoGame;
  status?: LottoStatus; // pending=未開獎、won=中獎、lost=未中、省略=全部
  page?: number;
  per_page?: number;
}

export interface LottoDraw {
  game: LottoGame;
  period: string; // 台彩期別字串，資訊用
  draw_date: string;
  main_nums: number[]; // 一般號/第一區（已排序）
  special: number; // 大樂透=特別號；威力彩=第二區號
}

export interface LottoDrawParams {
  game?: LottoGame;
  limit?: number;
}
